[NCCL] Add option to run NCCL on high priority cuda stream

mingzhe09088 opened the pull request:
Summary: This diff adds an option for the process group NCCL backend to pick high priority cuda streams. It lets cuda driver to prioritize NCCL kernels when there are compute kernels waiting. Here is an explanation about high priority cuda streams: https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1ge2be9e9858849bf62ba4a8b66d1c3540 

Test Plan: to add

Differential Revision: D23404286

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -669,19 +669,34 @@ They are used in specifying strategies for reduction collectives, e.g.,
+           const std::shared_ptr<::c10d::Store>&,
+           int,
+           int,
+           ::c10d::ProcessGroupNCCL::Options>())
```
Since this is also exposed to Python, shall we add some Python tests?

mrshenli reviewed (commented):
Hey @mingzhe09088, thanks for adding this. Could you please add some more description to the PR summary to explain the benefits of using a high priority stream?

mingzhe09088 commented on test/distributed/test_distributed.py:
```diff
@@ -678,6 +690,27 @@ def test_broadcast_cuda(self):
+        "Only Nccl backend supports stream",
+    )
+    @skip_if_no_gpu
+    def test_nccl_high_priority_stream(self):
```
@mrshenli I added one test here. Do you have suggestions about other tests to add?

pritamdamania87 commented on test/distributed/test_distributed.py:
```diff
@@ -680,14 +693,35 @@ def test_broadcast(self):
 
+    @unittest.skipIf(
+        BACKEND != "nccl",
+        "Only Nccl backend supports stream",
```
nit: "Only NCCL backend supports high priority stream"

pritamdamania87 commented on test/distributed/test_distributed.py:
```diff
@@ -680,14 +693,35 @@ def test_broadcast(self):
+        store = dist.PrefixStore(new_port, store)
+
+        opts = dist.ProcessGroupNCCL.Options()
+        opts.isHighPriority = False
```
Shouldn't we also test with True?

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -685,19 +685,34 @@ They are used in specifying strategies for reduction collectives, e.g.,
+
+  py::class_<::c10d::ProcessGroupNCCL::Options>(processGroupNCCL, "Options")
+      .def(py::init<>())
+      .def_readwrite("isHighPriority", &::c10d::ProcessGroupNCCL::Options::isHighPriority)
+      .def_readwrite("opTimeout", &::c10d::ProcessGroupNCCL::Options::opTimeout);
```
Don't we need some docs here for `isHighPriority` and `opTimeout` explaining what this means to users?

mingzhe09088 replied:
This will only be for power users. Not sure what's a good place to add the docs. Could you suggest a place for that?

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -685,19 +685,34 @@ They are used in specifying strategies for reduction collectives, e.g.,
+
+  py::class_<::c10d::ProcessGroupNCCL::Options>(processGroupNCCL, "Options")
+      .def(py::init<>())
+      .def_readwrite("isHighPriority", &::c10d::ProcessGroupNCCL::Options::isHighPriority)
+      .def_readwrite("opTimeout", &::c10d::ProcessGroupNCCL::Options::opTimeout);
```
In python land these should be snake_case right? (is_high_priority and op_timeout)

jiayisuse commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -143,6 +143,13 @@ class ProcessGroupNCCL : public ProcessGroup {
+    explicit Options();
+
+    std::chrono::milliseconds opTimeout;
+    bool isHighPriority;
```
Wondering if `isHighPriorityStream` is better

mingzhe09088 committed:
[NCCL] Add option to run NCCL on high priority cuda stream (#43796)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/43796

This diff adds an option for the process group NCCL backend to pick high priority cuda streams.

Test Plan: waitforsandcastle

Reviewed By: jiayisuse

Differential Revision: D23404286

fbshipit-source-id: 412f8216678c74d932f8143040809108d03eda79
